pp/go/relabeler/head/wal.go

```diff
@@ -218,36 +219,40 @@ func (d DecodedSegment) SampleCount() uint32 {
 	return d.sampleCount
 }
 
-func ReadSegment(reader io.Reader) (decodedSegment DecodedSegment, n int, err error) {
+func ReadSegment(reader io.Reader, decodedSegment *DecodedSegment) (n int, err error) {
 	br := &byteReader{r: reader}
 	var size uint64
 	size, err = binary.ReadUvarint(br)
 	if err != nil {
-		return decodedSegment, br.n, fmt.Errorf("failed to read segment size: %w", err)
+		return br.n, fmt.Errorf("failed to read segment size: %w", err)
 	}
 
 	crc32HashU64, err := binary.ReadUvarint(br)
 	if err != nil {
-		return decodedSegment, br.n, fmt.Errorf("failed to read segment crc32 hash: %w", err)
+		return br.n, fmt.Errorf("failed to read segment crc32 hash: %w", err)
 	}
 	crc32Hash := uint32(crc32HashU64)
 
 	sampleCountU64, err := binary.ReadUvarint(br)
 	if err != nil {
-		return decodedSegment, br.n, fmt.Errorf("failed to read segment sample count: %w", err)
+		return br.n, fmt.Errorf("failed to read segment sample count: %w", err)
 	}
 	decodedSegment.sampleCount = uint32(sampleCountU64)
 
-	decodedSegment.data = make([]byte, size)
+	if growTo := int(size) - len(decodedSegment.data); growTo > 0 {
```

Comment:
maybe
```suggestion
	if growTo := int(size) - cap(decodedSegment.data); growTo > 0 {
```
?

But slices.Grow checks capacity inside, so we can skip this check here, 'cause Grow will be inlined

Reply:
No, it's correct. `slices.Grow` do `n -= cap(s) - len(s)`
For example: we have slice with len 5 and capacity 6. We need slice with capacity 10.
We do `slice.Grow(10 - 5)` and `slice.Grow` increment capacity to 5 -= 6 - 5 = 4.

Reply:
I've looked at code of slices.Grow. I think it is better to just code here something like
```go
if int(size) > cap(decodedSegment.data) {
	decodedSegment.data = make([]byte, size)
} else {
	decodedSegment.data = decodedSegment.data[:size]
}
```